File: src/main/java/de/unistuttgart/iste/rss/tutormanagementsystem/model/scheincriteria/SheetTotalCriteria.java
package de.unistuttgart.iste.rss.tutormanagementsystem.model.scheincriteria;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import javax.persistence.Entity;
import javax.persistence.Transient;
import de.unistuttgart.iste.rss.tutormanagementsystem.formgenerator.annotations.IsScheinCriteria;
import de.unistuttgart.iste.rss.tutormanagementsystem.model.rating.Sheet;
import de.unistuttgart.iste.rss.tutormanagementsystem.model.student.Student;
import de.unistuttgart.iste.rss.tutormanagementsystem.service.BeanService;
import de.unistuttgart.iste.rss.tutormanagementsystem.service.SheetService;
import lombok.AccessLevel;
import lombok.Getter;

/**
 * SheetTotalCriteria
 */
@Getter
@Entity
@IsScheinCriteria(identifier = "sheetTotal")
public class SheetTotalCriteria extends PossiblePercentageCriteria {

    @Transient
    @Getter(value = AccessLevel.NONE)
    private SheetService sheetService = BeanService.getBean(SheetService.class);

    @Override
    public boolean isPassed(Student student) {
        double obtainedPointsOnAllSheets = 0d, maxObtainablePoints = 0d;

        for (Sheet sheet : sheetService.getAllSheets()) {
            if (!sheet.isBonusSheet()) {
                obtainedPointsOnAllSheets += student.getPoints(sheet);
                maxObtainablePoints += sheet.calculateMaxPoints();
            } else {
                obtainedPointsOnAllSheets += student.getPoints(sheet);
            }
        }

        if (this.isPercentage()) {
            return Double.compare(obtainedPointsOnAllSheets / maxObtainablePoints,
                    this.getValueNeeded()) >= 0;
        } else {
            return Double.compare(obtainedPointsOnAllSheets, this.getValueNeeded()) >= 0;
        }
    }

    @Override
    public ScheinCriteriaStatusResponseDTO getStatusDTO(Student student) {

        Map<UUID, ScheinCriteriaAdditionalStatusResponseDTO> infos = new HashMap<>();

        Double achieved = 0d;
        Double total = 0d;

        var sheets = sheetService.getAllSheets();

        for (Sheet sheet : sheets) {
            achieved += student.getPoints(sheet);
            total += sheet.calculateMaxPoints();

            infos.put(sheet.getId(), new ScheinCriteriaAdditionalStatusResponseDTO(achieved, total,
                    sheet.getSheetNo(), ScheinCriteriaUnit.POINT, PassedState.IGNORE));
        }
        return new ScheinCriteriaStatusResponseDTO(this.getId(), this.getIdentifier(),
                this.getName(), achieved, total, ScheinCriteriaUnit.POINT, this.isPassed(student));
    }

}

File: server/src/model/scheincriteria/criterias/SheetTotalCriteria.ts
import { PassedState, ScheinCriteriaUnit } from 'shared/dist/model/ScheinCriteria';
import { Student } from 'shared/dist/model/Student';
import scheincriteriaService from '../../../services/scheincriteria-service/ScheincriteriaService.class';
import sheetService from '../../../services/sheet-service/SheetService.class';
import { StatusCheckResponse } from '../Scheincriteria';
import {
  PossiblePercentageCriteria,
  possiblePercentageCriteriaSchema,
} from './PossiblePercentageCriteria';
import { Sheet } from 'shared/dist/model/Sheet';

export class SheetTotalCriteria extends PossiblePercentageCriteria {
  constructor(percentage: boolean, valueNeeded: number) {
    super('sheetTotal', percentage, valueNeeded);
  }

  async checkCriteriaStatus(student: Student): Promise<StatusCheckResponse> {
    const sheets = await sheetService.getAllSheets();
    const infos: StatusCheckResponse['infos'] = {};

    const { pointsAchieved, pointsTotal } = await this.checkAllSheets(sheets, student, infos);

    let passed: boolean = false;

    if (this.percentage) {
      passed = pointsAchieved / pointsTotal >= this.valueNeeded;
    } else {
      passed = pointsAchieved >= this.valueNeeded;
    }

    return {
      identifier: this.identifier,
      achieved: pointsAchieved,
      total: pointsTotal,
      passed,
      unit: ScheinCriteriaUnit.POINT,
      infos,
    };
  }

  private async checkAllSheets(
    sheets: Sheet[],
    student: Student,
    infos: StatusCheckResponse['infos']
  ) {
    let pointsAchieved = 0;
    let pointsTotal = 0;

    for (const sheet of sheets) {
      const achieved = await sheetService.getPointsOfStudent(student, sheet);
      const total = sheetService.getSheetTotalPoints(sheet);
      pointsAchieved += achieved;

      if (!sheet.bonusSheet) {
        pointsTotal += total;
      }

      infos[sheet.id] = {
        achieved,
        total,
        no: sheet.sheetNo,
        unit: ScheinCriteriaUnit.POINT,
        state: PassedState.IGNORE,
      };
    }

    return { pointsAchieved, pointsTotal };
  }
}

scheincriteriaService.registerBluePrint(
  new SheetTotalCriteria(false, 0),
  possiblePercentageCriteriaSchema
);
